include/dqlite.h

```diff
@@ -17,8 +17,278 @@
 
 int dqlite_version_number(void);
 
+/**
+ * Hold the value of a dqlite node ID. Guaranteed to be at least 64-bit long.
+ */
+typedef unsigned long long dqlite_node_id;
+
+typedef struct dqlite_server dqlite_server;
+typedef struct dqlite dqlite;
+typedef struct dqlite_stmt dqlite_stmt;
+
+/**
+ * Signature of a custom callback used to establish network connections
+ * to dqlite servers.
+ *
+ * @arg is a user data parameter, copied from the third argument of
+ * dqlite_server_set_connect_func. @addr is a (borrowed) abstract address
+ * string, as passed to dqlite_server_create or dqlite_server_set_peer. @fd is
+ * an address where a socket representing the connection should be stored. The
+ * callback should return zero if a connection was established successfully or
+ * nonzero if the attempt failed.
+ */
+typedef int (*dqlite_connect_func)(void *arg, const char *addr, int *fd);
+
+/* The following functions (dqlite_server_* and the functions that mimic the
+ * sqlite3 API) use SQLite result codes. dqlite tries to return result codes
+ * that match what SQLite would return in the same situation, but in some cases
+ * will return the less specific SQLITE_ERROR instead. Certain dqlite-specific
+ * error conditions have their own codes that are compatible with the SQLite
+ * set: */
+
+#define SQLITE_IOERR_NOT_LEADER (SQLITE_IOERR | (40 << 8))
+#define SQLITE_IOERR_LEADERSHIP_LOST (SQLITE_IOERR | (41 << 8))
+
+/* When some version of dqlite returns SQLITE_ERROR under certain
+ * circumstances, the developers reserve the right to switch to a more specific
+ * error code in future releases without announcing a breaking change. */
+
+/**
+ * Start configuring a dqlite server.
+ *
+ * The server will not start running until dqlite_server_start is called. @path
+ * is the path to a directory where the server (and attached client) will store
+ * its persistent state; the directory must exist. @addr is the abstract
+ * address of the server. A pointer to the new server object is stored in
+ * @server on success.
+ *
+ * The abstract address will be passed to the custom connect function configured
+ * with dqlite_server_set_connect_func, or to the default connect function if no
+ * custom function is configured. The default connect function accepts IPv4
+ * addressed, IPv6 addresses, and Unix-domain socket addresses; for a complete
+ * description of the syntax, see dqlite_server_set_bind_address.
+ *
+ * No reference to @path or @addr is kept after this function returns.
+ */
+int dqlite_server_create(const char *path,
+			 const char *addr,
```

Comment:
Any particular reason for which we are requiring the `addr` parameter?

I would expect that this bit of information is only needed for the very first time the server is started (similarly to what happens to other bits of information such as the peers), and will then be persisted in a file under `path` (along with some other info, such as an auto-generated server ID).

It seems we could just have:

```C
int dqlite_server_create(const char *path, **dqlite_server);
```

instead. And:

```C
int dqlite_server_set_address(dqlite_server* server, const char *address);
```

for the initial configuration, the very first time a server is started (it would be harmless to call `dqlite_server_set_address` unconditionally, like for the other functions).